### PR TITLE
PSyclone dependence analysis fix (closes #1315)

### LINKED ISSUE #1315: PSyclone is producing invalid code for LFRic
@TeranIvy has found that PSyclone master is producing incorrect LFRic code. This obviously needs to be fixed.

### PR DESCRIPTION
**Pull request by rupertford**
The code generation for LFRic is currently broken on master. This is because the data dependence analysis was including dependencies from all invokes, rather than just the current one.

### CODE REVIEW

**Comment by TeranIvy** [comment]
I ran the full developer and nightly test suites on Spice and XC40 for an LFRic branch that tidies up the algorithm layer for inconsistent declarations as otherwise the current PSyclone master reports errors, see [LFRic ticket #2737](https://code.metoffice.gov.uk/trac/lfric/ticket/2737). I built PSyclone test environment from this branch. I can confirm that the generated code for a sample of algorithms that use stencils is identical to the code generated with PSyclone 2.0.0 and that all test suites are green (that includes the KGO checks).

**Comment by rupertford** [comment]
Ready for first review from @arporter

File: src/psyclone/tests/psyir/nodes/node_test.py [hunk 1]
```diff
@@ -1086,3 +1088,48 @@ def test_node_constructor_with_parent():
...
+    # 1a: First node.
+    assert assign.following() == [
+        a_ref, multiply2, b_ref, multiply1, c_ref, d_ref]
+    assert not assign.preceding()
+
+    # 1b: Last node.
+    assert not d_ref.following()
+    assert d_ref.preceding() == [
+        assign, a_ref, multiply2, b_ref, multiply1, c_ref]
+    assert d_ref.preceding(reverse=True) == [
```
Comment (arporter) [inline src/psyclone/tests/psyir/nodes/node_test.py:1118]: It's not the fault of this PR but it seems strange to me that we bother supporting the `reverse` flag when all that is needed is a `.reverse()` on the result.

Comment (arporter) [reply src/psyclone/tests/psyir/nodes/node_test.py:1118]: (You can ignore this though.)

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good. Only request is that you extend the new test very slightly.
Changes are pycodestyle and pylint clean.
All tests and examples are OK with compilation
Documentation does not require updating.

File: src/psyclone/tests/psyir/nodes/node_test.py (hunk 1 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/nodes/node_test.py:1118]: I agree.

**Comment by rupertford** [comment]
Ready for next review from @arporter

**Review by arporter — APPROVED** [review]
Requested change has been made. Changes are pylint/pycodestyle clean.
All tests and examples are fine with compilation.
This is a small bug fix so no documentation changes are required.
Docs all still build OK.

**Pull request merged**